rstar/src/algorithm/iterators.rs

```diff
@@ -71,6 +71,62 @@ where
     }
 }
 
+pub struct SelectionWithDataIterator<'a, T, D, Func>
+where
+    T: RTreeObject + 'a,
+    Func: SelectionFunctionWithData<T, D>,
+{
+    func: Func,
+    current_nodes: SmallVec<[&'a RTreeNode<T>; 24]>,
+    _phantom: std::marker::PhantomData<D>,
+}
+
+impl<'a, T, D, Func> SelectionWithDataIterator<'a, T, D, Func>
+where
+    T: RTreeObject,
+    Func: SelectionFunctionWithData<T, D>,
+{
+    pub fn new(root: &'a ParentNode<T>, func: Func) -> Self {
+        let current_nodes = if func.should_unpack_parent(&root.envelope()) {
+            root.children.iter().collect()
+        } else {
+            SmallVec::new()
+        };
+
+        SelectionWithDataIterator {
+            func,
+            current_nodes,
+            _phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<'a, T, D, Func> Iterator for SelectionWithDataIterator<'a, T, D, Func>
+where
+    T: RTreeObject,
+    Func: SelectionFunctionWithData<T, D>,
+{
+    type Item = (&'a T, D);
+
+    fn next(&mut self) -> Option<(&'a T, D)> {
+        while let Some(next) = self.current_nodes.pop() {
+            match next {
+                RTreeNode::Leaf(ref t) => {
+                    if let Some(data) = self.func.should_unpack_leaf(t) {
```

Comment:
I wonder if we could reduce the amount of duplicate code if this would modify the existing `SelectIterator(Mut)` implementations and to always use `SelectionFunctionWithData`, just with a zero sized type like `()`?

Assuming a blanket impl of `SelectionFunctionWithData<T, ()>` for `SelectionFunction<T>`  is able to bridge between the two public API in a way that is backwards compatible?

(This would also provide a mutable variant of this query which I would suggest adding for completeness if above approach is not tenable.)

Reply:
I tried that approach first, but it required a ton of changes and I didn't really find it tenable. I'll add the mutable variant though.

Reply:
I don't mean to make this complicated but could expand on which intrusive changes were required?

I just tried the diff
```diff
diff --git a/rstar/src/algorithm/selection_functions.rs b/rstar/src/algorithm/selection_functions.rs
index 3aaeda5..9634b3f 100644
--- a/rstar/src/algorithm/selection_functions.rs
+++ b/rstar/src/algorithm/selection_functions.rs
@@ -53,6 +53,20 @@ where
     fn should_unpack_leaf(&self, _leaf: &T) -> Option<D>;
 }
 
+impl<F, T> SelectionFunctionWithData<T, ()> for F
+where
+    F: SelectionFunction<T>,
+    T: RTreeObject,
+{
+    fn should_unpack_parent(&self, envelope: &T::Envelope) -> bool {
+        SelectionFunction::should_unpack_parent(self, envelope)
+    }
+
+    fn should_unpack_leaf(&self, leaf: &T) -> Option<()> {
+        SelectionFunction::should_unpack_leaf(self, leaf).then(|| ())
+    }
+}
+
 pub struct SelectInEnvelopeFunction<T>
 where
     T: RTreeObject,
diff --git a/rstar/src/rtree.rs b/rstar/src/rtree.rs
index 726f4a6..a02a33d 100644
--- a/rstar/src/rtree.rs
+++ b/rstar/src/rtree.rs
@@ -376,7 +376,7 @@ where
         &self,
         selection_function: S,
     ) -> impl Iterator<Item = &T> {
-        SelectionIterator::new(&self.root, selection_function)
+        SelectionWithDataIterator::new(&self.root, selection_function).map(|(item, ())| item)
     }
 
     /// Mutable variant of [`locate_with_selection_function`](#method.locate_with_selection_function).
@@ -384,7 +384,7 @@ where
         &mut self,
         selection_function: S,
     ) -> impl Iterator<Item = &mut T> {
-        SelectionIteratorMut::new(&mut self.root, selection_function)
+        SelectionWithDataIteratorMut::new(&mut self.root, selection_function).map(|(item, ())| item)
     }
 
     /// Similar to [`locate_with_selection_function`](#method.locate_with_selection_function) but with
```
on top your branch and it does seem to build and test.

Is the problem the other explicitly named iterators based on `SelectionIterator{,Mut}`? Maybe writing a small named adaptor like `IgnoreData` that essentially does `.map(|(item, ())| item)` does so that those could be rewritten as e.g.
```rust
pub type LocateAllAtPoint<'a, T> = IgnoreData<SelectionIteratorWithData<'a, T, (), SelectAtPointFunction<T>>>;
```
could resolve this? Or to keep the public iterator type the same, reimplementing it as
```rust
pub struct SelectionIterator<'a, T, Func>(pub(crate) SelectionWithDataIterator<'a, T, (), Func>)
where
    T: RTreeObject + 'a,
    Func: SelectionFunction<T>;

impl<'a, T, Func> Iterator for SelectionIterator<'a, T, Func>
where
    T: RTreeObject,
    Func: SelectionFunction<T>,
{
    type Item = &'a T;

    fn next(&mut self) -> Option<&'a T> {
        self.0.next().map(|(item, ())| item)
    }
}
```
might be another option?

Reply:
Oh sorry, I think I misread. I'll try those changes.

Reply:
(Actually I think it would be simplest if we would demote all iterators returned by `RTree` methods to `impl Iterator<..>` but that should be a maintainer decision due to the implications for downstream code. We currently make those implications for some methods, so we should probably at least strive to be consistent, i.e. provide named types for all or none of the returns. cc @Stoeoef @urschrei)

Reply:
I don't have an immediate answer here – I'm in favour of consistency, though, and I'm fine with a breaking change (in my role as someone who's mostly facilitating maintenance of the crate, as opposed to developing it) if it's warranted. @rmanoka what do you think?

Reply:
It should be named types for all.  Hiding the type means we can't reference the iterator in other structures to be consumed elsewhere without incurring a dynamic dispatch or boxing it up, and on a performance focused library like this considering how it's used, that's important.

Reply:
@urschrei Sorry, I seem to have missed this comment.  I agree with @OvermindDL1 :  while we could expect the users to use generics if they wish to store the iterator, it makes such usage quite complicated.  Also, std. library seems to expose [most](https://doc.rust-lang.org/std/slice/struct.Iter.html) of [their](https://doc.rust-lang.org/std/iter/struct.Enumerate.html) iterator structures.

Reply:
@urschrei I also agree that iterators should be named. `impl Iterator` in return position is rarely useful for (most) libraries IMO.

Reply:
So I'm looking into your comments again now, @adamreichold. I'm kinda of the opinion that while we could reduce code duplication here by having everything use a `SelectionWithDataIterator`, it does make everything a bit more complicated. I'm not totally opposed to it, but I think there's an argument to be made for keeping things simple and duplicated as well.